desc/coils.py

```diff
@@ -1328,6 +1481,756 @@ def _check_type(coil0, coil):
+        return self._alpha_basis
+
+    @property
+    def alpha_N(self):
+        """Maximum mode number."""
+        return self._alpha_basis.N
+
+
+class _FiniteBuildCoil(_FramedCoil, Optimizable, ABC):
+    """Base class representing a magnetic field coil with finite build dimensions.
+
+    Implements the compute_self_field method.
+    Subclasses should inherit from this class as well as a subclass of Coil.
+
+    Parameters
+    ----------
+    cross_section_dims : array-like
+        Dimensions of the coil cross section, with 1 or 2 dimensions depending on
+        the cross section shape (circular or rectangular).
+        For circular cross sections, this should be a single value representing
+        the radius. For rectangular cross sections, this should be a 2-element array
+        representing the dimensions in the p and q directions, respectively.
+        Refer to fig 3 in https://arxiv.org/pdf/2310.12087 for the p and q directions.
+
+    cross_section_shape : str
+        Shape of the coil cross section, either 'circular' or 'rectangular'.
+    """
+
+    _io_attrs_ = _FramedCoil._io_attrs_ + [
+        "_cross_section_shape",
+        "_cross_section_dims",
+    ]
+
+    def __init__(self, cross_section_dims, cross_section_shape, *args, **kwargs):
+        if cross_section_shape == "circular":
+            assert (
+                len(cross_section_dims) == 1
+            ), "Circular cross section coils can only have one radius dimension"
+        elif cross_section_shape == "rectangular":
+            assert (
+                len(cross_section_dims) == 2
+            ), "Rectangular cross section coils can only have exactly two dimensions"
+        else:
+            raise ValueError(
+                "cross_section_shape must be 'circular' or 'rectangular', got "
+                f"{cross_section_shape}"
+            )
+        self._cross_section_shape = cross_section_shape
+        self._cross_section_dims = cross_section_dims
+        super().__init__(*args, **kwargs)
+
+    @optimizable_parameter
+    @property
+    def cross_section_dims(self):
+        """ndarray: Geometry of the coil, with cross section dimensions."""
+        return self._cross_section_dims
+
+    @cross_section_dims.setter
+    def cross_section_dims(self, new):
+        if self._cross_section_shape == "circular":
+            assert (
+                len(new) == 1
+            ), "Circular cross section coils can only have one radius dimension"
+        elif self._cross_section_shape == "rectangular":
+            assert (
+                len(new) == 2
+            ), "Rectangular cross section coils can only have exactly two dimensions"
+        self._cross_section_dims = new
+
+    @property
+    def cross_section_shape(self):
+        """str: Shape of the coil cross section, either 'circular' or 'rectangular'."""
+        return self._cross_section_shape
+
+    @classmethod
+    def prep_grid(cls, xsection_grid, centerline_grid=None):
```

Comment:
I'm not crazy about having two different grids for finite build coils that need to get combined in a particular way. I think it would make more sense to just require the user to give a 3d grid if they want 3d effects (and hence if they pass in a 1d grid it just does the usual centerline only stuff)